lib/datadog/tracing/span_operation.rb

```diff
@@ -435,18 +435,18 @@ def message
       # we don't want this SpanOperation to modify it further.
       def build_span
         Span.new(
-          @name && @name.dup,
+          @name.frozen? ? @name : @name.dup,
           duration: duration,
           end_time: @end_time,
           id: @id,
-          meta: meta && meta.dup,
-          metrics: metrics && metrics.dup,
+          meta: meta.dup,
+          metrics: metrics.dup,
           parent_id: @parent_id,
-          resource: @resource && @resource.dup,
-          service: @service && @service.dup,
+          resource: @resource.frozen? ? @resource : @resource.dup,
+          service: @service.frozen? ? @service : @service.dup,
           start_time: @start_time,
           status: @status,
-          type: @type && @type.dup,
+          type: @type.frozen? ? @type : @type.dup,
           trace_id: @trace_id
         )
       end
```

Comment:
The macOS Ruby 2.1 CI failure is not a flaky test -- https://github.com/DataDog/dd-trace-rb/runs/5118939789?check_suite_focus=true .

Today I learned that:
a) you can `nil.dup`, but only from Ruby 2.4+ (add one more to the "time wasted for old Rubies" bucket).
b) `nil` is only frozen from Ruby 2.2+
c) I can call `nil.freeze` for 2.1, and it seems to stick (!?!!)

which means that `foo.frozen? ? foo : foo.dup` works safely for 2.2+, but if a `nil` slips in there on 2.1 we may have a bad time. In particular, `@type` seems like it could be `nil` sometimes, looking at the rest of the class.

(I've taken the liberty of adding this PR to the old Ruby maintenance effort tracking sheet)

---

(Not blocking for this PR) I'm curious about the decision to dup instead of freeze the `meta` and `metrics`.

Given that the creation of a `Span` from a `SpanOperation` should be a terminal state (e.g. the `Span` will no longer be changed) -- could we just reuse the existing `meta` and `metrics` (and freeze them)? Leaving them mutable is kinda weird, since any changes done to them afterwards would not even propagate.

Reply:
> (Not blocking for this PR) I'm curious about the decision to dup instead of freeze the `meta` and `metrics`.
> 
> Given that the creation of a `Span` from a `SpanOperation` should be a terminal state (e.g. the `Span` will no longer be changed) -- could we just reuse the existing `meta` and `metrics` (and freeze them)? Leaving them mutable is kinda weird, since any changes done to them afterwards would not even propagate.

I was talking to @delner about this yesterday, I think it's a better option to make SpanOperation in some immutable state after `build_span`. I'll circle back to it.

Reply:
🤔 
```ruby
if RUBY_VERSION < '2.2'
  def nil.dup
    self
  end
end
```

Reply:
> I think it's a better option to make `SpanOperation` in some immutable state after `build_span`

This would be better; problem is a user code could hold a reference and attempt to mutate the `SpanOperation` after it has completed, resulting in an error. So this is a possible runtime hazard that didn't previously exist. I'd shy away from this, if possible.

Reply:
> This would be better; problem is a user code could hold a reference and attempt to mutate the SpanOperation after it has completed, resulting in an error. So this is a possible runtime error hazard that didn't previously exist. I'd shy away from this, if possible.

It would be an error when the operation requested (e.g. set_tag on a SpanOperation that was already flushed) will be silently ignored. Is silently ignoring set_tag better than an explicit error?

Reply:
> Is silently ignoring set_tag better than an explicit error?

Arguably yes. Raising exceptions could break user applications. I think technically the best way to handle it is issuing a one-time warning with the location from which the `set_tag` operation was performed. But this seems like polish for another milestone.